src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -62,45 +72,144 @@ class AdhocProcessResult:
 @dataclass(frozen=True)
 class ResolveExecutionDependenciesRequest:
     address: Address
-    execution_dependencies: tuple[str, ...] | None
+    execution_dependencies: tuple[str | runnable, ...] | None
     dependencies: tuple[str, ...] | None  # can go away after 2.17.0.dev0 per deprecation
 
 
 @dataclass(frozen=True)
 class ResolvedExecutionDependencies:
     digest: Digest
+    runnable_dependencies: RunnableDependencies | None
+
+
+@dataclass(frozen=True)
+class runnable:
```

Comment:
we could employ some clever AST transforms as a pre-parse step to automatically scope certain types..

Reply:
We never really resolved whether multiple fields or tagged values was the easiest way forward. We should probably resolve that first before we go down the rabbit hole.
Likely on a dedicated ticket instead of a PR

Reply:
Agree, just throwing around ideas :P

Reply:
We'll be doing with multiple fields to get this PR out the door (working on that as we speak), but I suspect tagged values will be the better long-term solution for this use case in particular.